Sources/swift-format/Subcommands/DumpConfiguration.swift

```diff
@@ -15,32 +15,59 @@ import Foundation
 import SwiftFormat
 
 extension SwiftFormatCommand {
-  /// Dumps the tool's default configuration in JSON format to standard output.
+  /// Dumps the tool's configuration in JSON format to standard output.
   struct DumpConfiguration: ParsableCommand {
     static var configuration = CommandConfiguration(
-      abstract: "Dump the default configuration in JSON format to standard output"
+      abstract: "Dump the configuration in JSON format to standard output",
+      discussion: """
+        Without any options, dumps the default configuration. When '--effective' is set, dumps the configuration that \
+        would be used if swift-format was executed from the current working directory (cwd), incorporating \
+        configuration files found in the cwd or its parents, or input from the '--configuration' option.
+        """
     )
 
+    /// Whether or not to dump the effective configuration.
+    @Flag(name: .shortAndLong, help: "Dump the effective instead of the default configuration.")
+    var effective: Bool = false
+
+    @OptionGroup()
+    var configurationOptions: ConfigurationOptions
+
+    func validate() throws {
+      if configurationOptions.configuration != nil && !effective {
+        throw ValidationError("'--configuration' is only valid in combination with '--effective'")
+      }
+    }
+
     func run() throws {
-      let configuration = Configuration()
-      do {
-        let encoder = JSONEncoder()
-        encoder.outputFormatting = [.prettyPrinted]
-        if #available(macOS 10.13, *) {
-          encoder.outputFormatting.insert(.sortedKeys)
-        }
+      let diagnosticPrinter = StderrDiagnosticPrinter(colorMode: .auto)
+      let diagnosticsEngine = DiagnosticsEngine(diagnosticsHandlers: [diagnosticPrinter.printDiagnostic])
+
+      let configuration: Configuration
+      if effective {
+        var configurationProvider = Frontend.ConfigurationProvider(diagnosticsEngine: diagnosticsEngine)
 
-        let data = try encoder.encode(configuration)
-        guard let jsonString = String(data: data, encoding: .utf8) else {
-          // This should never happen, but let's make sure we fail more gracefully than crashing, just
-          // in case.
-          throw FormatError(
-            message: "Could not dump the default configuration: the JSON was not valid UTF-8"
+        guard
+          let effectiveConfiguration = configurationProvider.provide(
+            forConfigPathOrString: configurationOptions.configuration,
+            orForSwiftFileAt: nil
           )
+        else {
+          // Already diagnosed in the called method through the diagnosticsEngine.
+          throw ExitCode.failure
         }
-        print(jsonString)
+
+        configuration = effectiveConfiguration
+      } else {
+        configuration = Configuration()
+      }
+
+      do {
+        let configurationAsJson = try configuration.asJsonString()
+        print(configurationAsJson)
```

Comment:
I guess this could be shortened slightly.

```suggestion
        print(try configuration.asJsonString())
```